src/messages/serialize.rs

```diff
@@ -0,0 +1,68 @@
+//! Serialization rules specified in [RFC-001#serializationdeserialization]
+//!
+//! We automatically serialize and deserialize using serde derivations where possible.
+//! Sometimes we need to implement ourselves, this file holds that code.
+//!
+//! [RFC-001#rules]: https://github.com/ZcashFoundation/redjubjub/blob/main/rfcs/0001-messages.md#serializationdeserialization
+
+use serde::ser::{Serialize, Serializer};
+
+use serde::de::{self, Deserialize, Deserializer, Visitor};
+
+use super::constants::{
+    AGGREGATOR_PARTICIPANT_ID, DEALER_PARTICIPANT_ID, MAX_SIGNER_PARTICIPANT_ID,
+};
+use super::*;
+
+use std::fmt;
+
```

Comment:
Isn't this 2 things guaranteed by the use of the `BTreeMap` ? I can't find the exact documentation but i made a quick test in https://github.com/ZcashFoundation/redjubjub/pull/114/commits/a066bb73dbb04795352aa4ab420249c991241ae8

Reply:
Oh, sorry, I think I wasn't very clear in what I meant :sweat:

Yeah, it's guaranteed by `BTreeMap`, so serialization from the `BTreeMap` will always work. Deserialization will also work, because inserting the elements will be _silently_ deduplicated and ordered by the `BTreeMap`.

What I wanted to say ask was if the "_silent_" handling by the `BTreeMap` is an issue. If the code receives a message that has an out of order or a repeated `ParticipantId`, should an error be produced? If so, then some `Deserialize` implementation will have to check for those inconsistencies and return an error :thinking:

Reply:
Right, I see what you mean now. I am unsure if we should handle it manually but lets don't do it now. I will leave this conversation unresolved until at least a ticket is created for this.